smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/HttpHeaderTraitValidator.java

```diff
@@ -68,15 +84,30 @@
 
         events.addAll(model.shapes(MemberShape.class)
                 .flatMap(member -> Trait.flatMapStream(member, HttpHeaderTrait.class))
-                .filter(pair -> BLACKLIST.contains(pair.getRight().getValue().toLowerCase(Locale.US)))
-                .map(pair -> danger(pair.getLeft(), String.format(
-                        "httpHeader cannot be set to `%s`", pair.getRight().getValue()
-                )))
+                .flatMap(pair -> OptionalUtils.stream(validateHeader(pair.left, pair.right)))
                 .collect(Collectors.toList()));
 
         return events;
     }
 
+    private Optional<ValidationEvent> validateHeader(MemberShape member, HttpHeaderTrait trait) {
+        String header = trait.getValue();
+
+        if (BLOCKLIST.contains(header.toLowerCase(Locale.ENGLISH))) {
+            return Optional.of(danger(member, trait, String.format(
+                    "`%s` is not an allowed HTTP header binding", header)));
+        }
+
+        for (int i = 0; i < header.length(); i++) {
+            if (!TCHAR.contains(header.charAt(i))) {
+                return Optional.of(danger(member, trait, String.format(
+                        "`%s` is not a valid HTTP header field name according to section 3.2 of RFC 7230", header)));
```

Comment:
Can you put the specific character that caused the failure in the error message?

Reply:
IMO, that isn't super important given we link to the RFC and show the header in the error message. I considered using a regular expression, but decided that this kind of check would be easier to understand. By keeping the validation message like this, we could theoretically switch to a regex and keep the same error message if we ever wanted to.